File: ValidParentheses/valid_parentheses.go
package ValidParentheses

func ValidParentheses(str string) bool {
	stack := make([]string, 0)
	pairs := map[string]string {
		"(": ")", 
		"{": "}", 
		"[": "]",
	}

	for _, v := range str {
		if _, ok := pairs[string(v)]; ok {
			stack = append(stack, string(v))
		} else {
			n := len(stack) - 1
			last := stack[n]
			if pairs[last] != string(v) {
				return false
			}
			stack = stack[:n]
		}
	}

	if len(stack) != 0 {
		return false
	}

	return true
}

File: ValidParentheses/valid_parentheses.rs
use std::collections::HashMap;

fn main() {
  let parentheses = "[(([{}]){[]})]{}".to_string();
  let is_valid = valid_parentheses(&parentheses);

  println!("{:?} is valid parentheses: {:?}", parentheses, is_valid);
}

fn valid_parentheses(str: &String) -> bool {
  let mut pairs = HashMap::new();
  let mut stack = Vec::new();

  pairs.insert("(".to_string(), ")".to_string());
  pairs.insert("{".to_string(), "}".to_string());
  pairs.insert("[".to_string(), "]".to_string());

  let str_char = str.chars();

  for (_, char) in str_char.enumerate() {
    match pairs.get(&char.to_string()) {
      Some(_) => {
        stack.push(char.to_string());
      },
      None => {
        match stack.pop() {
          Some(x) => {
            if pairs[&x] != char.to_string() {
              return false;
            }
          },
          _ => return false,
        }
      },
    }
  }

  if stack.len() != 0 {
    return false;
  }

  return true;
}